### PR TITLE
Allow leafletProxy arguments to contain JS()

### LINKED ISSUE #420: iconCreateFunction does not work when used in leafletProxy
see: http://stackoverflow.com/questions/43483696/iconcreatefunction-does-not-work-when-adding-markers-with-leafletproxy

basically when using markerClusterOptions(), if you try to pass your own JS to alter appearance of cluster, this only works when rendering leaflet() but breaks when later adding markers with leafletProxy()

### LINKED ISSUE #440: iconCreateFunction does NOT work when adding Markers with leafletProxy
https://stackoverflow.com/questions/43483696/iconcreatefunction-does-not-work-when-adding-markers-with-leafletproxy

### PR DESCRIPTION
**Pull request by jcheng5**
Fixes #420
Fixes #440

- [x] Make tests pass
- [x] Remove dependency on unexported function in htmlwidgets

htmlwidgets provides built-in support for the [`JS` function](https://rdrr.io/cran/htmlwidgets/man/JS.html), which lets you mark widget data string values in R to be evaluated as JS code when the widget data is deserialized in the browser. This isn't supported natively in Shiny though, so `leafletProxy` did not automatically inherit this behavior. The changes in this PR reimplement that mechanism for `leafletProxy`.

## Minimal reproducible example

From [this SO post](https://stackoverflow.com/questions/47922951/does-markercluster-work-together-with-leafletproxy-and-option-iconcreatefuncti) by @NilsOle, the following app should show the same behavior regardless of using leafletProxy or renderLeaflet. After clicking either button, zoom out to see clusters.

```r
library(shiny)
library(dplyr)
library(leaflet)

ui <- fluidPage(
  titlePanel("Hello Shiny!"),
  sidebarLayout(
    sidebarPanel(
      actionButton(inputId = "my_button1",
                   label = "Use leafletProxy()"),
      actionButton(inputId = "my_button2",
                   label = "Use renderLeaflet()")
    ),
    mainPanel(
      leafletOutput(
        outputId = "map",
        width = "100%",
        height = "300px"
      )
    )
  )
)

server <- function(input, output, session) {

  some_data <- data.frame(
    "lon"=c(4.905167,4.906357,4.905831),
    "lat"=c(52.37712,52.37783,52.37755),
    "number_var"=c(5,9,7),
    "name"=c("Jane","Harold","Mike"),
    stringsAsFactors = F
  )

  output$map <- renderLeaflet({
    return(
      leaflet(data = some_data[0,]) %>%
         addProviderTiles(providers$CartoDB.Positron) %>%
        fitBounds(
          min(some_data$lon),
          min(some_data$lat),
          max(some_data$lon),
          max(some_data$lat)
        ) %>%
        addMarkers(
          lng = ~lon,
          lat = ~lat,
          clusterOptions = markerClusterOptions(
            iconCreateFunction = JS(paste0("function(cluster) {",
                                           "return new L.DivIcon({",
                                           "html: '<div style=\"background-color:rgba(77,77,77,0.5)\"><span>' + cluster.getChildCount() + '</div><span>',",
                                           "className: 'marker-cluster'",
                                           "});",
                                           "}"))



          )
        )
    )
  })

  observeEvent(input$my_button1,{
      leafletProxy(mapId = "map",
                   session = session,
                   data = some_data) %>%
        addProviderTiles(providers$CartoDB.Positron) %>%
        clearMarkerClusters() %>%
        clearMarkers() %>%
        fitBounds(
          min(some_data$lon),
          min(some_data$lat),
          max(some_data$lon),
          max(some_data$lat)
        ) %>%
        addMarkers(
          lng = ~lon,
          lat = ~lat,
          clusterOptions = markerClusterOptions(
            iconCreateFunction = JS(paste0("function(cluster) {",
                                           "console.log('Here comes cluster',cluster); ",
                                           "return new L.DivIcon({",
                                           "html: '<div style=\"background-color:rgba(77,77,77,0.5)\"><span>' + cluster.getChildCount() + '</div><span>',",
                                           "className: 'marker-cluster'",
                                           "});",
                                           "}"))
          )
        )
  })

  observeEvent(input$my_button2,{
    output$map <- renderLeaflet({

      leaflet(data = some_data) %>%
        addProviderTiles(providers$CartoDB.Positron) %>%
        fitBounds(
          min(some_data$lon),
          min(some_data$lat),
          max(some_data$lon),
          max(some_data$lat)
        ) %>%
        addMarkers(
          lng = ~lon,
          lat = ~lat,
          clusterOptions = markerClusterOptions(
            iconCreateFunction = JS(paste0("function(cluster) {",
                                           "console.log('Here comes cluster',cluster); ",
                                           "return new L.DivIcon({",
                                           "html: '<div style=\"background-color:rgba(77,77,77,0.5)\"><span>' + cluster.getChildCount() + '</div><span>',",
                                           "className: 'marker-cluster'",
                                           "});",
                                           "}"))
          )
        )
    })
  })
}

shinyApp(ui = ui, server = server)
```

PR task list:
- [x] Update NEWS
- [x] Add tests (where appropriate)
  - R code tests: `tests/testthat/`
  - Visual tests: `R/zzz_viztest.R`
- [x] Update documentation with `devtools::document()`

### CODE REVIEW

**Comment by jcheng5** [comment]
Before merging, add a unit test that includes `JS()` on a leafletProxy call.

File: javascript/src/index.js [hunk 1]
```diff
@@ -280,6 +280,16 @@ HTMLWidgets.widget({
   }
 });
 
+function unpackArgs(arg) {
+  if (!arg.hasOwnProperty("arg") && !arg.hasOwnProperty("evals")) {
+    throw new Error("Malformed argument; .arg and .evals expected");
```
Comment (schloerke) [inline javascript/src/index.js:285]:
```suggestion
    throw new Error("Malformed argument; `arg` and `evals` properties expected");
```

File: javascript/src/index.js [hunk 2]
```diff
@@ -280,6 +280,16 @@ HTMLWidgets.widget({
   }
 });
 
+function unpackArgs(arg) {
+  if (!arg.hasOwnProperty("arg") && !arg.hasOwnProperty("evals")) {
```
Comment (schloerke) [inline javascript/src/index.js:284]:
```suggestion
  if (!arg.hasOwnProperty("value") && !arg.hasOwnProperty("evals")) {
```

**Review by schloerke — APPROVED** [review]

**Review by schloerke — APPROVED** [review]

**Comment by nicocriscuolo** [comment]
@jcheng5 I saw that the previous issues have been closed and everything was redirected to this one. What should be the working create custom cluster icons with JS() and make everything work in leafletProxy()? With the code above markers starts clustering when I zoom out, but there is no icon for the clusters.

Thanks again really a lot for working on solving this!

**Comment by jcheng5** [comment]
@nicocriscuolo Just to confirm--you tried it after running `remotes::install_github("rstudio/leaflet")` and restarting?

**Comment by nicocriscuolo** [comment]
@jcheng5 after installing this leaflet version everything works perfectly. Thanks really a lot for solving this!

**Comment by futluz** [comment]
@jcheng5...i tried this.... remotes::install_github("rstudio/leaflet")  - unfortunately didn't work for me. Would be grateful for any sugggestions'

**Pull request merged**